### PR TITLE
AWS lambda handlers added to route the requests to appropriate activity handlers

### PR DESCRIPTION
**Pull request by sandeepmjay**
Changes include:

- conversion of xapicaliper npm package utility to a microservice.
- Adding AWS Lambda handlers and routing requests to appropriate activity handlers for statement generation.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+      });
+      break;
+
+    // When no statement generator type match is found, an error message is relayed as a response.
+    default:
+      context.fail(new Error('Unrecognized statement generator request type : "' + statementGenerator + '"'));
+  }
+}
+
+// Since the error handler for all of the requests are the same. Creating a resuable callback method.
```
Comment (johncrossman) [inline index.js]: @sandeepmjay Typo: resuable

File: index.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+var resultHandler = function(context, err, statement, done) {
+  if(err) {
```
Comment (johncrossman) [inline index.js]: @sandeepmjay space needed after `if`

File: index.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    // AWS Lambda does not accept error objects sent in response body unless it is stringified.
+    // Additionally, RegEx can be set up on API Gateway to parse context object and relay custom error messages.
+    console.error('Statement generation failed. \n' + JSON.stringify(err, null, 4));
+    context.fail('Statement generation failed. \n' + JSON.stringify(err, null, 4));
+  } else {
+    console.log('Statement generated successfully !' + JSON.stringify(statement, null, 4));
+    context.succeed('Statement generated successfully ! \n' + JSON.stringify(statement, null, 4));
```
Comment (johncrossman) [inline index.js]: @sandeepmjay stray space after "successfully" in two lines above.

File: lib/statements/course.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+      'type': CaliperEntityType.COURSE_OFFERING,
+      'name': statement.metadata.name,
+      'description': statement.metadata.description
+    }, {
+      'http://id.tincanapi.com/extension/planned-start-time': StatementUtil.generateDate(statement.metadata.start),
+      'http://id.tincanapi.com/extension/planned-duration': StatementUtil.generateDate(statement.metadata.end)
+    })
+  }
+}, callback);
+return callback();
```
Comment (johncrossman) [inline lib/statements/course.js]: @sandeepmjay indentation fix needed on block above.

File: lib/util.js [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+  'number': joi.number(),
+  'date': joi.date(),
+  'email': joi.string().email(),
+  'uri': joi.string().uri(),
+  'array': joi.array(),
+  'agent': joi.object().keys({
+    'name': joi.string().optional(),
+    'created': joi.date().optional(),
+    'updated': joi.date().optional(),
+    'id': joi.string().uri().optional(),
```
Comment (johncrossman) [inline lib/util.js]: @sandeepmjay `id` is optional?

File: lib/util.js [hunk 6]
```diff
@@ -0,0 +1,608 @@
...
+ * @param  {Object}         [opts.XAPI.context.parent]        The properties that describe the parent of the learning activity
+ * @param  {Function}       callback                          Standard callback function
+ * @param  {Object}         callback.err                      An error that occured, if any
+ * @param  {Object}         callback.statement                The generated xAPI statement
+ */
+XAPI.processStatement = function(config, statement, opts, callback) {
+  var xapiStatement = XAPI.generateStatement(config, statement, opts);
+    XAPI.storeStatement(config, xapiStatement, function(err) {
+      return callback(null, xapiStatement);
+    });
```
Comment (johncrossman) [inline lib/util.js]: @sandeepmjay indentation fix needed above

Comment (sandeepmjay) [inline lib/util.js]: Fixed

File: lib/statements/course.js (hunk 4 above)
Comment (sandeepmjay) [inline lib/statements/course.js]: Fixed.

File: index.js (hunk 3 above)
Comment (sandeepmjay) [inline index.js]: Fixed.

File: index.js (hunk 2 above)
Comment (sandeepmjay) [inline index.js]: Fixed.

File: index.js (hunk 1 above)
Comment (sandeepmjay) [inline index.js]: Fixed.

File: lib/util.js (hunk 5 above)
Comment (sandeepmjay) [inline lib/util.js]: Yes. It depends on the activity type. For example, if it is course create activity,  id is not available on the user.

**Comment by johncrossman** [comment]
My comments are in. Minor fixes needed. Because we're establishing a new repo I'd suggest @paulkerschen give a quick glance.

**Comment by sandeepmjay** [comment]
Thanks @johncrossman. @paulkerschen: Now, you are it ! :)

File: index.js [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+// Since the error handler for all of the requests are the same. Creating a resuable callback method.
+var resultHandler = function(context, err, statement, done) {
+  if(err) {
+    // AWS Lambda does not accept error objects sent in response body unless it is stringified.
+    // Additionally, RegEx can be set up on API Gateway to parse context object and relay custom error messages.
+    console.error('Statement generation failed. \n' + JSON.stringify(err, null, 4));
+    context.fail('Statement generation failed. \n' + JSON.stringify(err, null, 4));
+  } else {
+    console.log('Statement generated successfully !' + JSON.stringify(statement, null, 4));
```
Comment (paulkerschen) [inline index.js]: @sandeepmjay Newline missing in this log statement.

File: index.js [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+// Since the error handler for all of the requests are the same. Creating a reusable callback method.
+var resultHandler = function(context, err, statement, done) {
+
+  if(err) {
+    // AWS Lambda does not accept error objects sent in response body unless it is stringified.
+    // Additionally, RegEx can be set up on API Gateway to parse context object and relay custom error messages.
+    console.error('Statement generation failed. \n' + JSON.stringify(err, null, 4));
+    context.fail('Statement generation failed. \n' + JSON.stringify(err, null, 4));
+  } else {
+    console.log('Statement generated successfully!' + JSON.stringify(statement, null, 4));
```
Comment (paulkerschen) [inline index.js]: @sandeepmjay Newline character missing in log statement above.

File: tests/actor.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+// TODO
+
+// Identified group
+
+// xAPI
+
+// Caliper
+
+// TODO
```
Comment (paulkerschen) [inline tests/actor.js]: @sandeepmjay Newline missing at end of file.

File: package.json [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    "caliperjs": "git://github.com/IMSGlobal/caliper-js-public.git#public",
+    "config" : "latest",
+    "joi": "latest",
+    "lodash": "latest",
+    "request": "latest",
+    "tincanjs": "latest",
+    "uuid": "latest"
+  },
+  "engines": {
+    "node": ">=0.10"
```
Comment (paulkerschen) [inline package.json]: @sandeepmjay the Node version running under Lambda will be much greater, correct?

File: package.json [hunk 11]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "xapicaliper",
+  "description": "Easily convert learning activities to xAPI and Caliper statements",
+  "version": "0.3.0",
+  "repository": {
+    "type": "git",
+    "url": "git://github.com/nicolaasmatthijs/xapicaliper"
```
Comment (paulkerschen) [inline package.json]: @sandeepmjay repo URL should be updated

File: lib/util.js [hunk 12]
```diff
@@ -0,0 +1,610 @@
+var _ = require('lodash');
+var crypto = require('crypto');
+var joi = require('joi');
+var request = require('request');
+var tincan = require('tincanjs');
+var uuid = require('uuid');
+
+
```
Comment (paulkerschen) [inline lib/util.js]: @sandeepmjay Extra linebreak

File: lib/util.js [hunk 13]
```diff
@@ -0,0 +1,610 @@
...
+ * @param  {Boolean}        metadataRules[metadataField].required   Whether the metadata field is required
+ * @return {Object}                                                 A validation error that occured, if any
+ */
+var validate = module.exports.validate = function(config, statement, metadataRules) {
+  var validationSchema = {};
+
+  // Verify that the required platform values are present
+  validationSchema.platform = {
+    'url': VALIDATORS.uri.required(),
+    'name':VALIDATORS.string.required()
```
Comment (paulkerschen) [inline lib/util.js]: @sandeepmjay Space missing after colon

File: lib/util.js [hunk 14]
```diff
@@ -0,0 +1,610 @@
...
+    'platform': config.platform,
+    'timestamp': statement.timestamp,
+    'actor': statement.actor,
+    'metadata': metadata
+  }, joi.object().keys(validationSchema));
+
+  if (validationResult.error) {
+    console.log(validationResult.error.details);
+    return {'code': 400, 'msg': validationResult.error.details[0].message};
+
```
Comment (paulkerschen) [inline lib/util.js]: @sandeepmjay Extra linebreak.

File: lib/statements/course.js [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+      'type': 'date',
+      'required': false
+    }
+  });
+  if (validationError) {
+    console.log("Error propagated :" + JSON.stringify(validationError));
+    return callback(validationError);
+  }
+
+
```
Comment (paulkerschen) [inline lib/statements/course.js]: @sandeepmjay Extra linebreak.

File: index.js (hunk 8 above)
Comment (sandeepmjay) [inline index.js]: Fixed

File: package.json (hunk 11 above)
Comment (sandeepmjay) [inline package.json]: changed it with the repo in ets-berkeley

File: package.json (hunk 10 above)
Comment (sandeepmjay) [inline package.json]: switched to 4.3.2. Good catch on package.json !

File: lib/util.js (hunk 13 above)
Comment (sandeepmjay) [inline lib/util.js]: Fixed

File: lib/util.js (hunk 12 above)
Comment (sandeepmjay) [inline lib/util.js]: Fixed.

File: lib/util.js (hunk 14 above)
Comment (sandeepmjay) [inline lib/util.js]: Fixed

File: lib/statements/course.js (hunk 15 above)
Comment (sandeepmjay) [inline lib/statements/course.js]: Fixed.

File: tests/actor.js (hunk 9 above)
Comment (sandeepmjay) [inline tests/actor.js]: Not sure about this.

**Comment by sandeepmjay** [comment]
@johncrossman @paulkerschen - Thanks for the review. Amended the styling elements. Merge at your convenience.

**Comment by paulkerschen** [comment]
Thanks, @sandeepmjay!

**Pull request merged**